workflow/scripts/modify_prenetwork.py

```diff
@@ -679,6 +679,111 @@ def aladin_mobility_demand(n):
 
 
+def add_hydrogen_turbines(n):
+    """
+    This adds links that instead of a gas turbine use a hydrogen turbine.
+
+    It is assumed that the efficiency stays the same. This function is
+    only applied to German nodes.
+    """
+    logger.info("Adding hydrogen turbine technologies for Germany.")
+
+    gas_carrier = ["OCGT", "CCGT"]
+    for carrier in gas_carrier:
+        gas_plants = n.links[
+            (n.links.carrier == carrier)
+            & (n.links.index.str[:2] == "DE")
+            & (n.links.p_nom_extendable)
+        ].index
+        if gas_plants.empty:
+            continue
+        h2_plants = n.links.loc[gas_plants].copy()
+        h2_plants.carrier = h2_plants.carrier.str.replace(carrier, "H2 " + carrier)
+        h2_plants.index = h2_plants.index.str.replace(carrier, "H2 " + carrier)
+        h2_plants.bus0 = h2_plants.bus0.str.replace("gas", "H2")
+        h2_plants.bus2 = ""
+        h2_plants.efficiency2 = 0
+        # add the new links
+        n.import_components_from_dataframe(h2_plants, "Link")
+
+    # special handling of CHPs
+    gas_plants = n.links[
+        (n.links.carrier == "urban central gas CHP")
+        & (n.links.index.str[:2] == "DE")
+        & (n.links.p_nom_extendable)
+    ].index
+    h2_plants = n.links.loc[gas_plants].copy()
+    h2_plants.carrier = h2_plants.carrier.str.replace("gas", "H2")
+    h2_plants.index = h2_plants.index.str.replace("gas", "H2")
+    h2_plants.bus0 = h2_plants.bus0.str.replace("gas", "H2")
+    h2_plants.bus3 = ""
+    h2_plants.efficiency3 = 0
+    n.import_components_from_dataframe(h2_plants, "Link")
+
+
+def force_retrofit(n, params):
+    """
+    This function forces the retrofit of gas turbines from params["force"] on.
+
+    Extendable gas links are deleted.
+    """
+
+    logger.info("Forcing retrofit of gas turbines to hydrogen turbines.")
+
+    gas_carrier = ["OCGT", "CCGT", "urban central gas CHP"]
+    # deleting extendable gas turbine plants
+    to_drop = n.links[
+        (n.links.carrier.isin(gas_carrier))
+        & (n.links.p_nom_extendable)
+        & (n.links.index.str[:2] == "DE")
+    ].index
+    n.links.drop(to_drop, inplace=True)
+
+    # forcing retrofit
+    for carrier in ["OCGT", "CCGT"]:
+        gas_plants = n.links[
+            (n.links.carrier == carrier)
+            & (n.links.index.str[:2] == "DE")
+            & (n.links.build_year >= params["start"])
+        ].index
+        if gas_plants.empty:
+            continue
+
+        h2_plants = n.links.loc[gas_plants].copy()
+        h2_plants.carrier = h2_plants.carrier.str.replace(
+            carrier, "H2 retro " + carrier
+        )
+        h2_plants.index = h2_plants.index.str.replace(carrier, "H2 retro " + carrier)
+        h2_plants.bus0 = h2_plants.bus0.str.replace("gas", "H2")
+        h2_plants.bus2 = ""
+        h2_plants.efficiency -= params["efficiency_loss"]
+        h2_plants.efficiency2 = 1  # default value
```

Comment:
it's a bit strange that the default value is 0 for H2 plants, but 1 for H2 retro plants